arrow/src/row/mod.rs

```diff
@@ -0,0 +1,577 @@
+///
+/// A valid value is encoded as a big endian length, with the most significant bit set, followed
+/// by the byte values
+///
+/// ## Dictionary Encoding
+///
+/// **Not currently implemented**
+///
+/// Dictionaries are materialized to their values in the row format, this may require dramatically
+/// more memory than the source [`Array`]. It is recommended that the batch size is kept
+/// sufficiently small that this doesn't cause issues
+///
+/// # Ordering
+///
+/// ## Float Ordering
+///
+/// Floats are totally ordered in accordance to the `totalOrder` predicate as defined
+/// in the IEEE 754 (2008 revision) floating point standard.
+///
+/// The ordering established by this does not always agree with the
+/// [`PartialOrd`] and [`PartialEq`] implementations of `f32`. For example,
+/// they consider negative and positive zero equal, while this does not
+///
+/// ## Null Ordering
+///
+/// The row format currently orders nulls as less than non-null values. A future extension
+/// could allow configuring this, by inverting the representation of the null bit
+///
+/// ## Reverse Column Ordering
+///
+/// The row format does not currently support reversing the ordering of a specific column. A
+/// future extension could allow configuring this by inverting the bit representation of values
+/// for the column in question
+///
+/// ## Reconstruction
+///
+/// Given a schema it would theoretically be possible to reconstruct the columnar data from
+/// the row format, however, this is currently not supported. It is recommended that the row
+/// format is instead used to obtain a sorted list of row indices, which can then be used
+/// with [`take`]:[crate::compute::take] to obtain a sorted [`RecordBatch`]
+///
+/// [`memcmp`]:[https://www.man7.org/linux/man-pages/man3/memcmp.3.html]
+///
+#[derive(Debug)]
+pub struct RowBatch {
+    buffer: Box<[u8]>,
+    offsets: Box<[usize]>,
+}
+
+impl RowBatch {
+    /// Create a [`RowBatch`] from the provided [`RecordBatch`]
+    pub fn new(batch: &RecordBatch) -> Self {
+        let lengths = get_row_lengths(batch);
+        let mut offsets = Vec::with_capacity(batch.num_rows() + 1);
+        offsets.push(0);
+
+        let mut cur_offset = 0_usize;
+        for l in lengths {
+            offsets.push(cur_offset);
+            cur_offset = cur_offset.checked_add(l).expect("overflow");
+        }
+
+        let buffer = vec![0_u8; cur_offset];
+
+        let mut out = Self {
+            buffer: buffer.into(),
+            offsets: offsets.into(),
+        };
+
+        for col in batch.columns() {
+            // We encode a column at a time to minimise dispatch overheads
+            encode_column(col, &mut out)
+        }
+
+        if cfg!(debug_assertions) {
+            assert_eq!(*out.offsets.last().unwrap(), out.buffer.len());
+            out.offsets
+                .windows(2)
+                .for_each(|w| assert!(w[0] <= w[1], "offsets should be monotonic"));
```

Comment:
```suggestion
                .for_each(|w| assert!(w[0] < w[1], "offsets should be monotonic"));
```

They should be strictly increasing, right?